src/torchmetrics/functional/regression/kendall.py

```diff
@@ -0,0 +1,396 @@
+        target_ties_p2,
+        n_total,
+    ) = _get_metric_metadata(preds, target, variant)
+    con_min_dis_pairs = concordant_pairs - discordant_pairs
+
+    tau = _calculate_tau(
+        preds, target, concordant_pairs, discordant_pairs, con_min_dis_pairs, n_total, preds_ties, target_ties, variant
+    )
+    p_value = (
+        _calculate_p_value(
+            con_min_dis_pairs,
+            n_total,
+            preds_ties,
+            preds_ties_p1,
+            preds_ties_p2,
+            target_ties,
+            target_ties_p1,
+            target_ties_p2,
+            variant,
+            alternative,
+        )
+        if alternative
+        else None
+    )
+
+    # Squeeze tensor if num_outputs=1
+    if tau.shape[0] == 1:
+        tau = tau.squeeze()
+        p_value = p_value.squeeze() if p_value is not None else None
+
+    return tau.clamp(-1, 1), p_value
+
+
+def kendall_rank_corrcoef(
+    preds: Tensor,
+    target: Tensor,
+    variant: Literal["a", "b", "c"] = "b",
+    t_test: bool = False,
+    alternative: Optional[Literal["two-sided", "less", "greater"]] = "two-sided",
+) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+    r"""Computes `Kendall Rank Correlation Coefficient`_.
+
+    .. math::
+        tau_b = \frac{C - D}{\sqrt{(C + D + T_{preds}) * (C + D + T_{target})}}
+
+        tau_c = 2 * \frac{C - D}{n ** 2 * \frac{m - 1}{m}}
+
+    where :math:`C` is represents concordant pairs, :math:`D` stands for discordant pairs and :math:`T` represents
+    a total number of ties. Definition according to `The Treatment of Ties in Ranking Problems`_.
+
+    Args:
+        preds: Ordered sequence of data
+        target: Ordered sequence of data
+        variant: Indication of which variant of Kendall's tau to be used
+        t_test: Indication whether to run t-test
+        alternative: Alternative hypothesis for t-test. Possible values:
+            - 'two-sided': the rank correlation is nonzero
+            - 'less': the rank correlation is negative (less than zero)
+            - 'greater':  the rank correlation is positive (greater than zero)
+
+    Return:
+        Correlation tau statistic
+        (Optional) p-value of corresponding statistical test (asymptotic)
+
+    Raises:
+        ValueError: If ``t_test`` is not of a type bool
+        ValueError: If ``t_test=True`` and ``alternative=None``
+
+    Example (single output regression):
+        >>> from torchmetrics.functional.regression import kendall_rank_corrcoef
+        >>> target = torch.tensor([3, -0.5, 2, 1])
+        >>> preds = torch.tensor([2.5, 0.0, 2, 8])
+        >>> kendall_rank_corrcoef(preds, target)
+        tensor(0.3333)
+
+    Example (multi output regression):
+        >>> from torchmetrics.functional.regression import kendall_rank_corrcoef
+        >>> target = torch.tensor([[3, -0.5], [2, 1]])
+        >>> preds = torch.tensor([[2.5, 0.0], [2, 8]])
+        >>> kendall_rank_corrcoef(preds, target)
+        tensor([1., 1.])
```

Comment:
could we include an third example with `t_test=True` to show that the metric can output an second value